rampwf/prediction_types/detection.py

```diff
@@ -3,56 +3,96 @@
-    self.check_y_pred_dimensions()
+        if index_list is None:  # we combine the full list
+            index_list = range(len(predictions_list))
+        y_comb_list = [predictions_list[i].y_pred for i in index_list]
 
+        matches = []
 
-def _check_y_pred_dimensions(self):
-    # XXX should check that prediction is an array of lists or Nones.
-    pass
+        for a, b in itertools.combinations(y_comb_list, 2):
+            idx1, idx2, ious = _match_tuples([(x, y, r) for (c, x, y, r) in a],
+                                             [(x, y, r) for (c, x, y, r) in b])
 
+            idx1 = idx1[ious > 0.5]
+            idx2 = idx2[ious > 0.5]
 
-@classmethod
-def _combine(cls, predictions_list, index_list=None):
-    if len(predictions_list) == 0:
-        return []
-    else:
-        return predictions_list[0]
+            for i1, i2 in zip(idx1, idx2):
+                combined = (np.asarray(a[i1]) + np.array(b[i2])) / 2
+                matches.append(combined)
 
+        combined = _greedy_nms(matches)
 
-def _set_valid_in_train(self, predictions, test_is):
-    """Set a cross-validation slice."""
-    self.y_pred[test_is] = predictions.y_pred
+        combined_predictions = cls(y_pred=combined)
+        return combined_predictions
 
+    def set_valid_in_train(self, predictions, test_is):
+        """Set a cross-validation slice."""
+        self.y_pred[test_is] = predictions.y_pred
 
-@property
-def _valid_indexes(self):
-    """Return valid indices (e.g., a cross-validation slice)."""
-    return self.y_pred != np.empty(len(self.y_pred), dtype=np.object)
+    @property
+    def valid_indexes(self):
+        """Return valid indices (e.g., a cross-validation slice)."""
+        return self.y_pred != np.empty(len(self.y_pred), dtype=np.object)
 
 
-def make_detection(label_names=[]):
-    Predictions = type(
-        'Predictions',
-        (BasePrediction,),
-        {'__init__': _detection_init,
-         'check_y_pred_dimensions': _check_y_pred_dimensions,
-         'combine': _combine,
-         'set_valid_in_train': _set_valid_in_train,
-         'valid_indexes': _valid_indexes,
-         })
+def make_detection():
     return Predictions
+
+
+def _greedy_nms(y_pred, iou_threshold=0.45):
+    y_pred = np.asarray(y_pred)
+
+    boxes_left = np.copy(y_pred)
+    # This is where we store the boxes that make it through the non-maximum suppression
+    maxima = []
+
+    # While there are still boxes left to compare...
+    while boxes_left.shape[0] > 0:
+        # ...get the index of the next box with the highest confidence...
+        maximum_index = np.argmax(boxes_left[:, 0])
+        # ...copy that box and...
+        maximum_box = np.copy(boxes_left[maximum_index])
+        # ...append it to `maxima` because we'll definitely keep it
+        maxima.append(maximum_box)
+        # Now remove the maximum box from `boxes_left`
+        boxes_left = np.delete(boxes_left, maximum_index, axis=0)
```

Comment:
Just an idea while reading this: it may be simpler to have a list of boxes sorted by confidence and then pop them one by one.

Reply:
I just copy pasted this code from somewhere .. :-) so probably can be improved yes. But didn't look in detail up to now.